activesupport/lib/active_support/descendants_tracker.rb

```diff
@@ -6,21 +6,82 @@ module ActiveSupport
   # This module provides an internal implementation to track descendants
   # which is faster than iterating through ObjectSpace.
   module DescendantsTracker
+    # DescendantsArray is an array that contains weak references to classes.
+    class DescendantsArray # :nodoc:
+      include Enumerable
+
+      def initialize
+        @refs = []
+      end
+
+      def initialize_copy(orig)
+        @refs = @refs.dup
+      end
+
+      def <<(klass)
+        cleanup!
+        @refs << WeakRef.new(klass)
+      end
+
+      def each
+        @refs.each do |ref|
+          yield ref.__getobj__
+        rescue WeakRef::RefError
+        end
+      end
+
+      def refs_size
+        @refs.size
+      end
+
+      def cleanup!
+        @refs.delete_if { |ref| !ref.weakref_alive? }
+      end
+
+      def reject!
+        @refs.reject! do |ref|
+          yield ref.__getobj__
+        rescue WeakRef::RefError
+          true
+        end
+      end
+    end
+
     @@direct_descendants = {}
+    @@descendants_preloaded = DescendantsArray.new
```

Comment:
Ideally this would be a `Set`, but since it need to contain WeakRef, it meant implementing a custom container etc.

Given that I don't expect this array to contain that much entry, and that it's for development use only, it didn't seem really worth it to me.